File: web/src/types/backtest.ts
export type { KlinePayload } from "@/types/public"

export interface BacktestStartRequest {
  run_id?: string | null
  symbols?: string[] | null
  start_ts?: number | null
  end_ts?: number | null
  initial_balance?: number | null
  fee_bps?: number | null
  slippage_bps?: number | null
  ai_model_id?: string | null
  prompt_variant?: string | null
  btc_eth_leverage?: number | null
  altcoin_leverage?: number | null
  interval?: string | null
  decision_every?: number | null
}

export interface BacktestRunIdRequest {
  run_id: string
}

export interface BacktestLabelRequest {
  run_id: string
  label: string
}

export interface BacktestQueryParams {
  run_id?: string | null
  limit?: number | null
}

export interface KlinesQuery {
  symbol: string
  interval?: string | null
  limit?: number | null
}

export interface BacktestRunActionPayload {
  run_id: string
  message: string
}

export interface BacktestMessagePayload {
  message: string
}

export interface BacktestRunConfigPayload {
  run_id?: string
  symbols?: string[]
  start_ts?: number
  end_ts?: number
  initial_balance?: number
  fee_bps?: number
  slippage_bps?: number
  ai_model_id?: string
  prompt_variant?: string
  btc_eth_leverage?: number
  altcoin_leverage?: number
  interval?: string
  decision_every?: number
}

export interface BacktestRunSummaryPayload {
  equity_last?: number
  final_equity?: number
  initial_balance?: number
  max_drawdown_pct?: number
  total_trades?: number
  winning_trades?: number
}

export interface BacktestRunPayload {
  run_id: string
  label: string
  state: string
  last_error: string
  summary: BacktestRunSummaryPayload
  created_at: number
  updated_at: number
}

export interface BacktestStatusItemPayload extends BacktestRunPayload {
  config: BacktestRunConfigPayload
}

export interface BacktestEquityPointPayload {
  ts: number
  equity: number
  available: number
  pnl: number
  pnl_pct: number
  dd_pct: number
  cycle: number
}

export interface BacktestTradePayload {
  id: string
  ts: number
  symbol: string
  action: string
  side: string
  qty: number
  price: number
  fee: number
  realized_pnl: number
  leverage: number
  cycle: number
  liquidation: boolean
}

export interface BacktestDecisionPayload {
  id: string
  ts: number
  symbol: string
  decision: string
  confidence: number
  reason: string
  cycle: number
}

export interface BacktestMetricsItemPayload {
  run_id: string
  state: string
  initial_balance: number
  final_equity: number
  total_pnl: number
  total_pnl_pct: number
  total_trades: number
  winning_trades: number
  win_rate_pct: number
  max_drawdown_pct: number
}

export interface BacktestStatusPayload {
  status: BacktestStatusItemPayload
}

export interface BacktestRunsPayload {
  runs: BacktestRunPayload[]
  count: number
}

export interface BacktestEquityPayload {
  points: BacktestEquityPointPayload[]
  count: number
}

export interface BacktestTradesPayload {
  trades: BacktestTradePayload[]
  count: number
}

export interface BacktestTracePayload {
  trace: BacktestDecisionPayload[]
}

export interface BacktestDecisionsPayload {
  decisions: BacktestDecisionPayload[]
  count: number
}

export interface BacktestMetricsPayload {
  metrics: BacktestMetricsItemPayload
}

export interface BacktestExportPayload {
  run_id: string
  trades: BacktestTradePayload[]
  equity: BacktestEquityPointPayload[]
  exported_at: number
}

File: core/src/contracts/backtest.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

pub use crate::contracts::public::KlinePayload;

#[derive(Debug, Deserialize)]
pub struct BacktestStartRequest {
    pub run_id: Option<String>,
    pub symbols: Option<Vec<String>>,
    pub start_ts: Option<i64>,
    pub end_ts: Option<i64>,
    pub initial_balance: Option<f64>,
    pub fee_bps: Option<f64>,
    pub slippage_bps: Option<f64>,
    pub ai_model_id: Option<String>,
    pub prompt_variant: Option<String>,
    pub btc_eth_leverage: Option<i64>,
    pub altcoin_leverage: Option<i64>,
    pub interval: Option<String>,
    pub decision_every: Option<usize>,
}

#[derive(Debug, Deserialize)]
pub struct BacktestRunIdRequest {
    pub run_id: String,
}

#[derive(Debug, Deserialize)]
pub struct BacktestLabelRequest {
    pub run_id: String,
    pub label: String,
}

#[derive(Debug, Deserialize)]
pub struct BacktestQueryParams {
    pub run_id: Option<String>,
    pub limit: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct KlinesQuery {
    pub symbol: String,
    pub interval: Option<String>,
    pub limit: Option<i64>,
}

#[derive(Debug, Clone, Serialize)]
pub struct BacktestRunActionPayload {
    pub run_id: String,
    pub message: &'static str,
}

#[derive(Debug, Clone, Serialize)]
pub struct BacktestMessagePayload {
    pub message: &'static str,
}

#[derive(Debug, Clone, Serialize)]
pub struct BacktestStatusPayload {
    pub status: Value,
}

#[derive(Debug, Clone, Serialize)]
pub struct BacktestRunsPayload {
    pub runs: Vec<Value>,
    pub count: usize,
}

#[derive(Debug, Clone, Serialize)]
pub struct BacktestEquityPayload {
    pub points: Vec<Value>,
    pub count: usize,
}

#[derive(Debug, Clone, Serialize)]
pub struct BacktestTradesPayload {
    pub trades: Vec<Value>,
    pub count: usize,
}

#[derive(Debug, Clone, Serialize)]
pub struct BacktestTracePayload {
    pub trace: Vec<Value>,
}

#[derive(Debug, Clone, Serialize)]
pub struct BacktestDecisionsPayload {
    pub decisions: Vec<Value>,
    pub count: usize,
}

#[derive(Debug, Clone, Serialize)]
pub struct BacktestMetricsPayload {
    pub metrics: Value,
}

#[derive(Debug, Clone, Serialize)]
pub struct BacktestExportPayload {
    pub run_id: String,
    pub trades: Vec<Value>,
    pub equity: Vec<Value>,
    pub exported_at: i64,
}
